pmlc/dialect/pxa/ir/interfaces.td

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020, Intel Corporation
+
+#ifndef __PMLC_PXA_INTERFACES__
+#define __PMLC_PXA_INTERFACES__
+
+include "mlir/IR/OpBase.td"
+
+class PxaOpInterface<string name> : OpInterface<name>;
+
+def PxaReadOpInterface : PxaOpInterface<"PxaReadOpInterface"> {
+  let description = [{
+      Interface to query characteristics of read-like ops with affine
+      restrictions.
+  }];
+
+  let methods = [
+    InterfaceMethod<
+      /*desc=*/"Returns the memref operand to read from.",
+      /*retTy=*/"::mlir::Value",
+      /*methodName=*/"getMemRef",
+      /*args=*/(ins),
+      /*methodBody*/[{}],
+      /*defaultImplementation=*/ [{
+        ConcreteOp op = cast<ConcreteOp>(this->getOperation());
+        return op.getOperand(op.getMemRefOperandIndex());
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/"Returns the type of the memref operand.",
+      /*retTy=*/"::mlir::MemRefType",
+      /*methodName=*/"getMemRefType",
+      /*args=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        ConcreteOp op = cast<ConcreteOp>(this->getOperation());
+        return op.getMemRef().getType().template cast<::mlir::MemRefType>();
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/"Returns affine map operands.",
+      /*retTy=*/"::mlir::Operation::operand_range",
+      /*methodName=*/"getMapOperands",
+      /*args=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        ConcreteOp op = cast<ConcreteOp>(this->getOperation());
+        return llvm::drop_begin(op.getOperands(), 1);
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/[{
+        Returns the affine map used to index the memref for this operation.
+      }],
+      /*retTy=*/"::mlir::AffineMap",
+      /*methodName=*/"getAffineMap",
+      /*args=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        ConcreteOp op = cast<ConcreteOp>(this->getOperation());
+        return op.getAffineMapAttr().getValue();
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/"Returns the AffineMapAttr associated with 'memref'.",
+      /*retTy=*/"::mlir::NamedAttribute",
+      /*methodName=*/"getAffineMapAttrForMemRef",
+      /*args=*/(ins "::mlir::Value":$memref),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        ConcreteOp op = cast<ConcreteOp>(this->getOperation());
+        assert(memref == getMemRef());
+        return {Identifier::get(op.getMapAttrName(), op.getContext()),
+                op.getAffineMapAttr()};
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/"Returns the value read by this operation.",
+      /*retTy=*/"::mlir::Value",
+      /*methodName=*/"getValue",
+      /*args=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        return cast<ConcreteOp>(this->getOperation());
+      }]
+    >,
+  ];
+}
+
+def PxaWriteOpInterface : PxaOpInterface<"PxaWriteOpInterface"> {
```

Comment:
Should we call this PxaReduceOpInterface?  Write doesn't imply the fact that a reduction is happening (unless type is assign).